src/Interpreters/Cache/FileSegment.cpp

```diff
@@ -917,10 +924,11 @@ void FileSegment::use()
     }
 }
 
-FileSegments::iterator FileSegmentsHolder::completeAndPopFrontImpl()
+FileSegmentsHolder::FileSegmentsHolder(FileSegments && file_segments_, bool complete_on_dtor_)
+    : file_segments(std::move(file_segments_)), complete_on_dtor(complete_on_dtor_)
 {
-    front().complete();
-    return file_segments.erase(file_segments.begin());
+    CurrentMetrics::add(CurrentMetrics::FilesystemCacheHoldFileSegments, file_segments.size());
```

Comment:
Why do we need both profile event and metric?

Reply:
Because they serve different purposes